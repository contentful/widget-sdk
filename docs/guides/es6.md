Writing ES6
===========

Angular services can no can be written using ES6 syntax and are transpiled with
[Babel][babel].

ES6 transpilation is enabled for files with the `.es6.js` extension. See the
[ES6 Modules](#es6-modules) section for more info on how to integrate ES6 code
with Angular.

We use a restricted set of Babel transform plugins that we will extend when
necessary. See `tools/app-babel-options.js` for the list of supported plugins.

ES6 syntax is not yet supported for directives and controllers.

[babel]: http://babeljs.io/


Adding a Transform Plugin
-------------------------

To add support for an ES6 feature one needs to adjust the Babel configuration
and the ESLint configuration. The babel configuration is located in
`tools/app-babel-options.js`

Be sure to prefer the `loose` option for plugins if available. This will create
less correct code but it will be more portable without using polyfills.


Linting
-------

We have to different ESLint configurations `src/javascripts/.eslint-es5.yml` and
`src/javascripts/.eslint-es6.yml` for ES5 and ES6 files, repectively.

The ES6 configuration excludes ES6 syntax that is not transformed through the
[`no-restricted-syntax` rule][no-restricted-syntax-rule]. If you add a Babel
plugin make sure to also update the ESLint configuration.

We use two shell scripts `bin/lint-all` and `bin/lint-file` to lint against the
appropriate configuraiton. You can use the latter to integrate ESLint into your
code editor or IDE.

[no-restricted-syntax-rule]: http://eslint.org/docs/rules/no-restricted-syntax


ES6 Modules
-----------

The ES6 module syntax (using the keywords `import` and `export`) is transpiled
into code that is then plugged into Angularâ€™s DI module system.

Assume the file `src/javascripts/a/b/x.es6.js` has the following export
~~~js
export default const def = 'default'
export const foo = 'foo'
~~~
This file is transformed into an Angular service with its name `a/b/x` derived
from the file name. Thus, from Angular we can import this file as follows.
~~~js
var X = $injector.get('a/b/x')
assert.equal(X.default, 'default')
assert.equal(X.foo, 'foo')
~~~

Conversely, we can import Angular services in ES6 modules.
~~~js
import * as X from 'x'
~~~
is equivalent to
~~~
var X = $injector.get('x')
~~~

We also support relative imports. Assume the file `src/javascripts/a/b/x.es6.js`
has the following import.
~~~js
import * as Y from '../y'
import * as Z from './z'
~~~
Imports are resolved relative to the module ID (which coincides with the service
name) of the file they are used in. Since the file above has the module ID
`a/b/c` this is equivalent to
~~~js
var Y = $injector.get('a/y')
var Z = $injector.get('a/b/z')
~~~

The implementation uses the [SystemJS Babel plugin][babel-systemjs] and is
provided in `src/javascripts/prelude.js`.

[babel-systemjs]: http://babeljs.io/docs/plugins/transform-es2015-modules-systemjs/
