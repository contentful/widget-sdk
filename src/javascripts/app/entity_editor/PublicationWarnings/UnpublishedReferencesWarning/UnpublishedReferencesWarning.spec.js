import React from 'react';
import Enzyme from 'enzyme';
import { create as createPublicationManager } from '../index.es6';
import { registerUnpublishedReferencesWarning as createRegisterUnpublishedReferencesWarning } from './index.es6';
import Confirm from './UnpublishedReferencesConfirm.es6';

const mockStubs = {
  modalOpenStub: jest.fn().mockResolvedValue(true)
};

jest.mock(
  'app/common/ModalLauncher.es6',
  () => ({
    open: (...args) => mockStubs.modalOpenStub(...args)
  }),
  { virtual: true }
);

describe('UnpublishedReferencesWarning', () => {
  const createField = name => ({
    name: name,
    locale: 'en-US'
  });

  const createEntry = () => ({
    sys: {
      type: 'Entry',
      id: 1
    }
  });

  const createAsset = () => ({
    sys: {
      type: 'Asset',
      id: 1
    }
  });
  describe('registerUnpublishedReferencesWarning', () => {
    let publicationWarningManager, registerUnpublishedReferencesWarning;
    beforeEach(() => {
      publicationWarningManager = createPublicationManager();
      registerUnpublishedReferencesWarning = createRegisterUnpublishedReferencesWarning(
        publicationWarningManager
      );
    });
    afterEach(() => {});
    it('does not show confirmation if no unpublished references provided', async () => {
      registerUnpublishedReferencesWarning({
        getData: async () => {
          return Promise.resolve({
            field: createField(),
            references: []
          });
        }
      });

      await expect(publicationWarningManager.show()).toResolve();
      expect(mockStubs.modalOpenStub).not.toBeCalled();
    });

    it('shows confirm if unpublished references provided', async () => {
      registerUnpublishedReferencesWarning({
        getData: async () => {
          return Promise.resolve({
            field: createField(),
            references: [createEntry()]
          });
        }
      });

      await expect(publicationWarningManager.show()).toResolve();
      expect(mockStubs.modalOpenStub).toBeCalledTimes(1);
    });
  });

  describe('UnpublishedReferencesConfirm', () => {
    const defaultProps = () => ({
      isShown: true,
      onConfirm: jest.fn(),
      onCancel: jest.fn(),
      unpublishedReferencesInfo: []
    });

    it('renders confirmation dialog', () => {
      const wrapper = Enzyme.shallow(<Confirm {...defaultProps()} />);

      expect(wrapper).toMatchSnapshot();
    });

    describe.each([
      ['This entry links to 1 unpublished entry', createEntry()],
      ['This entry links to 2 unpublished entries', createEntry(), createEntry()],
      ['This entry links to 1 unpublished asset', createAsset()],
      ['This entry links to 2 unpublished assets', createAsset(), createAsset()],
      [
        'This entry links to 1 unpublished entry and 1 unpublished asset',
        createEntry(),
        createAsset()
      ],
      [
        'This entry links to 1 unpublished entry and 2 unpublished assets',
        createEntry(),
        createAsset(),
        createAsset()
      ],
      [
        'This entry links to 2 unpublished entries and 1 unpublished asset',
        createEntry(),
        createEntry(),
        createAsset()
      ],
      [
        'This entry links to 2 unpublished entries and 2 unpublished assets',
        createEntry(),
        createEntry(),
        createAsset(),
        createAsset()
      ]
    ])('renders following title "%s"', (title, ...references) => {
      const wrapper = Enzyme.mount(
        <Confirm
          {...defaultProps()}
          unpublishedReferencesInfo={[
            { field: createField('field-with-unpublished-entities'), references }
          ]}
        />
      );

      expect(wrapper.find('[data-test-id="cf-ui-modal-header"]').text()).toBe(title);
    });

    it('renders list of fields with unpublished references', () => {
      const wrapper = Enzyme.mount(
        <Confirm
          {...defaultProps()}
          unpublishedReferencesInfo={[
            {
              field: createField('field-with-unpublished-entry'),
              references: [createEntry()]
            },
            {
              field: createField('field-with-unpublished-entries'),
              references: [createEntry(), createEntry()]
            },
            {
              field: createField('field-with-unpublished-asset'),
              references: [createAsset()]
            },
            {
              field: createField('field-with-unpublished-assets'),
              references: [createAsset(), createAsset()]
            },
            {
              field: createField('field-with-unpublished-mixed'),
              references: [createEntry(), createAsset()]
            },
            {
              field: createField('field-with-unpublished-mixed'),
              references: [createEntry(), createEntry(), createAsset(), createAsset()]
            }
          ]}
        />
      );

      expect(wrapper.find('[data-test-id="cf-ui-modal-content"]')).toMatchSnapshot();
    });
  });
});
