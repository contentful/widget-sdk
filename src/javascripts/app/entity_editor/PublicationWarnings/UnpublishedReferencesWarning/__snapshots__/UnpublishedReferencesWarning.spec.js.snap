// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnpublishedReferencesWarning UnpublishedReferencesConfirm renders confirmation dialog 1`] = `
<ModalConfirm
  cancelLabel="Cancel"
  cancelTestId="cf-ui-modal-confirm-cancel-button"
  confirmLabel="Got it, publish anyway"
  confirmTestId="cf-ui-modal-confirm-confirm-button"
  intent="positive"
  isConfirmDisabled={false}
  isConfirmLoading={false}
  isShown={true}
  onCancel={[MockFunction]}
  onConfirm={[MockFunction]}
  shouldCloseOnEscapePress={true}
  shouldCloseOnOverlayClick={true}
  size="medium"
  testId="unpublished-refs-confirm"
  title="This entry links to "
>
  <p>
    It appears that you’ve linked to 
     that 
    hasn't
     
    been published yet.
  </p>
  <ul />
</ModalConfirm>
`;

exports[`UnpublishedReferencesWarning UnpublishedReferencesConfirm renders list of fields with unpublished references 1`] = `
<div
  className="ModalContent__ModalContent___2mf3h"
  data-test-id="cf-ui-modal-content"
>
  <p>
    It appears that you’ve linked to 
    6 unpublished entries and 6 unpublished assets
     that 
    haven't
     
    been published yet.
  </p>
  <ul>
    <li
      key="0"
    >
      <strong>
        field-with-unpublished-entry (en-US)
      </strong>
       — 
      1 unpublished entry
    </li>
    <li
      key="1"
    >
      <strong>
        field-with-unpublished-entries (en-US)
      </strong>
       — 
      2 unpublished entries
    </li>
    <li
      key="2"
    >
      <strong>
        field-with-unpublished-asset (en-US)
      </strong>
       — 
      1 unpublished asset
    </li>
    <li
      key="3"
    >
      <strong>
        field-with-unpublished-assets (en-US)
      </strong>
       — 
      2 unpublished assets
    </li>
    <li
      key="4"
    >
      <strong>
        field-with-unpublished-mixed (en-US)
      </strong>
       — 
      1 unpublished entry and 1 unpublished asset
    </li>
    <li
      key="5"
    >
      <strong>
        field-with-unpublished-mixed (en-US)
      </strong>
       — 
      2 unpublished entries and 2 unpublished assets
    </li>
  </ul>
</div>
`;
